src/Microsoft.NetCore.Analyzers/Core/Security/JsonNetTypeNameHandling.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.NetCore.Analyzers.Security.Helpers;
+
+namespace Microsoft.NetCore.Analyzers.Security
+{
+    /// <summary>
+    /// For using a <see cref="T:Newtonsoft.Json.TypeNameHandling"/> values other than None.
+    /// </summary>
+    [SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "The comment references a type that is not referenced by this compilation.")]
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal class JsonNetTypeNameHandling : DiagnosticAnalyzer
+    {
+        // TODO paulming: Help links URLs.
+        internal static readonly DiagnosticDescriptor Rule =
+            SecurityHelpers.CreateDiagnosticDescriptor(
+                "CA2326",
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetTypeNameHandlingTitle),
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetTypeNameHandlingMessage),
+                isEnabledByDefault: false,
+                helpLinkUri: null,
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetTypeNameHandlingDescription));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create<DiagnosticDescriptor>(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            analysisContext.EnableConcurrentExecution();
+
+            analysisContext.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    WellKnownTypeProvider wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(
+                        compilationStartAnalysisContext.Compilation);
+
+                    if (!wellKnownTypeProvider.TryGetTypeByMetadataName(
+                            WellKnownTypeNames.NewtonsoftJsonTypeNameHandling,
+                            out INamedTypeSymbol typeNameHandlingSymbol))
+                    {
+                        return;
+                    }
+
+                    bool IsOtherThanNone(IFieldReferenceOperation fieldReferenceOperation)
```

Comment:
Nit: Move the local functions at the end of the method, and add an explicit return statement prior to local function.